File: GO/BenchmarkRunner.go
// benchmark_runner.go
package main

import (
    "fmt"
    "log"
    "os"
    "runtime"
    "time"
    "ordenacao-go/Algorithms"
    "ordenacao-go/Structures"
    "ordenacao-go/Utils"
)

const NUM_EXECUTIONS = 10

type BenchmarkRun struct {
    ExecutionNumber int
    TimeNs         int64
    TimeMicros     float64
    TimeMs         float64
    MemoryUsedKB   float64
    MemoryUsedMB   float64
}

type BenchmarkResult struct {
    StructureName  string
    StructureType  string
    DataSize       int
    Runs           []BenchmarkRun
    AvgTimeMs      float64
    AvgMemoryMB    float64
    MinTimeMs      float64
    MaxTimeMs      float64
    MinMemoryMB    float64
    MaxMemoryMB    float64
}

type SystemInfo struct {
    Timestamp  string
    OS         string
    Arch       string
    CPUs       int
    GoVersion  string
    TotalMemGB float64
}

func getMemStats() runtime.MemStats {
    var m runtime.MemStats
    runtime.ReadMemStats(&m)
    return m
}

func forceGC() {
    runtime.GC()
    runtime.GC()
    time.Sleep(100 * time.Millisecond)
}

func runSingleBenchmark(ds Utils.DataStructure, data []int) (timeNs int64, memoryBytes uint64) {
    forceGC()
    
    memBefore := getMemStats()
    startTime := time.Now()
    
    // Processo de ordenação
    ds.FromArray(data)
    arr := ds.ToArray()
    sorted := Algorithms.CountingSort(arr)
    ds.FromArray(sorted)
    
    duration := time.Since(startTime)
    memAfter := getMemStats()
    
    timeNs = duration.Nanoseconds()
    memoryBytes = memAfter.TotalAlloc - memBefore.TotalAlloc
    
    return
}

func benchmarkStructure(structName, structType string, createStructure func(int) Utils.DataStructure, data []int) BenchmarkResult {
    result := BenchmarkResult{
        StructureName: structName,
        StructureType: structType,
        DataSize:      len(data),
        Runs:          make([]BenchmarkRun, 0, NUM_EXECUTIONS),
    }
    
    fmt.Printf("  %s: ", structName)
    
    var totalTime float64
    var totalMemory float64
    minTime := float64(^uint64(0) >> 1)
    maxTime := 0.0
    minMemory := float64(^uint64(0) >> 1)
    maxMemory := 0.0
    
    for i := 0; i < NUM_EXECUTIONS; i++ {
        ds := createStructure(len(data))
        timeNs, memoryBytes := runSingleBenchmark(ds, data)
        
        run := BenchmarkRun{
            ExecutionNumber: i + 1,
            TimeNs:         timeNs,
            TimeMicros:     float64(timeNs) / 1000.0,
            TimeMs:         float64(timeNs) / 1e6,
            MemoryUsedKB:   float64(memoryBytes) / 1024.0,
            MemoryUsedMB:   float64(memoryBytes) / (1024.0 * 1024.0),
        }
        
        result.Runs = append(result.Runs, run)
        
        totalTime += run.TimeMs
        totalMemory += run.MemoryUsedMB
        
        if run.TimeMs < minTime {
            minTime = run.TimeMs
        }
        if run.TimeMs > maxTime {
            maxTime = run.TimeMs
        }
        if run.MemoryUsedMB < minMemory {
            minMemory = run.MemoryUsedMB
        }
        if run.MemoryUsedMB > maxMemory {
            maxMemory = run.MemoryUsedMB
        }
        
        fmt.Print(".")
    }
    
    result.AvgTimeMs = totalTime / float64(NUM_EXECUTIONS)
    result.AvgMemoryMB = totalMemory / float64(NUM_EXECUTIONS)
    result.MinTimeMs = minTime
    result.MaxTimeMs = maxTime
    result.MinMemoryMB = minMemory
    result.MaxMemoryMB = maxMemory
    
    fmt.Printf(" Concluído (Média: %.2fms, %.2fMB)\n", result.AvgTimeMs, result.AvgMemoryMB)
    
    return result
}

func runBenchmarksForSize(data []int, logFile *os.File) []BenchmarkResult {
    results := make([]BenchmarkResult, 0)
    
    fmt.Fprintf(logFile, "\n=== TAMANHO DO DATASET: %d elementos ===\n", len(data))
    fmt.Printf("\n--- Testando com %d elementos ---\n", len(data))
    
    // Lista Linear
    result := benchmarkStructure("Lista Linear", "linear", 
        func(size int) Utils.DataStructure { 
            return Structures.NewListLinear(size + 100) 
        }, data)
    results = append(results, result)
    logResult(logFile, result)
    
    // Lista Dinâmica
    result = benchmarkStructure("Lista Dinâmica", "dynamic",
        func(size int) Utils.DataStructure { 
            return Structures.NewListDynamic() 
        }, data)
    results = append(results, result)
    logResult(logFile, result)
    
    // Fila Linear
    result = benchmarkStructure("Fila Linear", "linear",
        func(size int) Utils.DataStructure { 
            return Structures.NewQueueLinear(size + 100) 
        }, data)
    results = append(results, result)
    logResult(logFile, result)
    
    // Fila Dinâmica
    result = benchmarkStructure("Fila Dinâmica", "dynamic",
        func(size int) Utils.DataStructure { 
            return Structures.NewQueueDynamic() 
        }, data)
    results = append(results, result)
    logResult(logFile, result)
    
    // Pilha Linear
    result = benchmarkStructure("Pilha Linear", "linear",
        func(size int) Utils.DataStructure { 
            return Structures.NewStackLinear(size + 100) 
        }, data)
    results = append(results, result)
    logResult(logFile, result)
    
    // Pilha Dinâmica
    result = benchmarkStructure("Pilha Dinâmica", "dynamic",
        func(size int) Utils.DataStructure { 
            return Structures.NewStackDynamic() 
        }, data)
    results = append(results, result)
    logResult(logFile, result)
    
    return results
}

func logResult(logFile *os.File, result BenchmarkResult) {
    fmt.Fprintf(logFile, "\n%s (%s) - %d elementos:\n", 
        result.StructureName, result.StructureType, result.DataSize)
    fmt.Fprintf(logFile, "----------------------------------------\n")
    
    // Cabeçalho da tabela
    fmt.Fprintf(logFile, "Execução | Tempo (ms) | Memória (MB)\n")
    fmt.Fprintf(logFile, "---------|------------|-------------\n")
    
    // Dados de cada execução
    for _, run := range result.Runs {
        fmt.Fprintf(logFile, "%8d | %10.2f | %11.2f\n", 
            run.ExecutionNumber, run.TimeMs, run.MemoryUsedMB)
    }
    
    fmt.Fprintf(logFile, "---------|------------|-------------\n")
    fmt.Fprintf(logFile, "MÉDIA    | %10.2f | %11.2f\n", 
        result.AvgTimeMs, result.AvgMemoryMB)
    fmt.Fprintf(logFile, "MÍNIMO   | %10.2f | %11.2f\n", 
        result.MinTimeMs, result.MinMemoryMB)
    fmt.Fprintf(logFile, "MÁXIMO   | %10.2f | %11.2f\n", 
        result.MaxTimeMs, result.MaxMemoryMB)
    fmt.Fprintf(logFile, "\n")
}

func getSystemInfo() SystemInfo {
    var m runtime.MemStats
    runtime.ReadMemStats(&m)
    
    return SystemInfo{
        Timestamp:  time.Now().Format("02/01/2006 15:04:05"),
        OS:         runtime.GOOS,
        Arch:       runtime.GOARCH,
        CPUs:       runtime.NumCPU(),
        GoVersion:  runtime.Version(),
        TotalMemGB: float64(m.Sys) / (1024 * 1024 * 1024),
    }
}

func logSystemInfo(logFile *os.File, info SystemInfo) {
    fmt.Fprintf(logFile, "==========================================\n")
    fmt.Fprintf(logFile, "     BENCHMARK - COUNTING SORT EM GO     \n")
    fmt.Fprintf(logFile, "==========================================\n")
    fmt.Fprintf(logFile, "Data/Hora: %s\n", info.Timestamp)
    fmt.Fprintf(logFile, "Sistema: %s/%s\n", info.OS, info.Arch)
    fmt.Fprintf(logFile, "CPUs: %d\n", info.CPUs)
    fmt.Fprintf(logFile, "Go Version: %s\n", info.GoVersion)
    fmt.Fprintf(logFile, "Memória Total Alocada: %.2f GB\n", info.TotalMemGB)
    fmt.Fprintf(logFile, "Algoritmo: Counting Sort\n")
    fmt.Fprintf(logFile, "Execuções por teste: %d\n", NUM_EXECUTIONS)
    fmt.Fprintf(logFile, "==========================================\n")
}

func createSummaryTable(logFile *os.File, allResults [][]BenchmarkResult) {
    fmt.Fprintf(logFile, "\n==========================================\n")
    fmt.Fprintf(logFile, "           TABELA RESUMO FINAL           \n")
    fmt.Fprintf(logFile, "==========================================\n\n")
    
    // Cabeçalho
    fmt.Fprintf(logFile, "%-20s | %-8s | ", "Estrutura", "Tipo")
    sizes := []int{100, 1000, 10000, 100000, 1000000}
    for _, size := range sizes {
        fmt.Fprintf(logFile, "%10d | ", size)
    }
    fmt.Fprintf(logFile, "\n")
    
    // Linha separadora
    fmt.Fprintf(logFile, "%-20s-+-%-8s-+-", "--------------------", "--------")
    for range sizes {
        fmt.Fprintf(logFile, "-----------+-")
    }
    fmt.Fprintf(logFile, "\n")
    
    // Estruturas na ordem
    Structures := []string{
        "Lista Linear", "Lista Dinâmica", 
        "Fila Linear", "Fila Dinâmica",
        "Pilha Linear", "Pilha Dinâmica",
    }
    
    // Dados de tempo
    fmt.Fprintf(logFile, "\nTEMPO MÉDIO (ms):\n")
    for _, structName := range Structures {
        for sizeIdx, results := range allResults {
            for _, result := range results {
                if result.StructureName == structName && sizeIdx == 0 {
                    fmt.Fprintf(logFile, "%-20s | %-8s | ", 
                        result.StructureName, result.StructureType)
                    break
                }
            }
        }
        
        for sizeIdx := range sizes {
            for _, result := range allResults[sizeIdx] {
                if result.StructureName == structName {
                    fmt.Fprintf(logFile, "%10.2f | ", result.AvgTimeMs)
                    break
                }
            }
        }
        fmt.Fprintf(logFile, "\n")
    }
    
    // Dados de memória
    fmt.Fprintf(logFile, "\nMEMÓRIA MÉDIA (MB):\n")
    for _, structName := range Structures {
        for sizeIdx, results := range allResults {
            for _, result := range results {
                if result.StructureName == structName && sizeIdx == 0 {
                    fmt.Fprintf(logFile, "%-20s | %-8s | ", 
                        result.StructureName, result.StructureType)
                    break
                }
            }
        }
        
        for sizeIdx := range sizes {
            for _, result := range allResults[sizeIdx] {
                if result.StructureName == structName {
                    fmt.Fprintf(logFile, "%10.2f | ", result.AvgMemoryMB)
                    break
                }
            }
        }
        fmt.Fprintf(logFile, "\n")
    }
}

func main() {
    // Criar arquivo de log
    logFileName := fmt.Sprintf("benchmark_%s.log", 
        time.Now().Format("20060102_150405"))
    logFile, err := os.Create(logFileName)
    if err != nil {
        log.Fatal("Erro ao criar arquivo de log:", err)
    }
    defer logFile.Close()
    
    // Log system info
    sysInfo := getSystemInfo()
    logSystemInfo(logFile, sysInfo)
    
    fmt.Println("=== BENCHMARK COUNTING SORT ===")
    fmt.Printf("Executando %d testes para cada estrutura...\n", NUM_EXECUTIONS)
    fmt.Printf("Salvando resultados em: %s\n", logFileName)
    
    // Tamanhos para testar
    sizes := []int{100, 1000, 10000, 100000, 1000000}
    allResults := make([][]BenchmarkResult, 0)
    
    for _, size := range sizes {
        // Carregar dados
        data, err := Utils.LoadRatingsCSV("ratings.csv", size)
        if err != nil {
            // Se não conseguir ler o arquivo, criar dados de teste
            fmt.Printf("Aviso: Usando dados aleatórios para tamanho %d\n", size)
            data = make([]int, size)
            for i := range data {
                // Gerar ratings aleatórios de 1 a 10 (equivalente a 0.5 a 5.0)
                data[i] = (i % 10) + 1
            }
        }
        
        results := runBenchmarksForSize(data, logFile)
        allResults = append(allResults, results)
    }
    
    // Criar tabela resumo
    createSummaryTable(logFile, allResults)
    
    fmt.Fprintf(logFile, "\n==========================================\n")
    fmt.Fprintf(logFile, "         BENCHMARK FINALIZADO            \n")
    fmt.Fprintf(logFile, "==========================================\n")
    
    fmt.Printf("\n✅ Benchmark concluído! Resultados salvos em: %s\n", logFileName)
}

File: PYTHON/benchmark_runner.py
import time
import gc
import sys
import platform
import tracemalloc
from datetime import datetime
from typing import List, Dict, Callable
from dataclasses import dataclass

# Adicionar diretórios ao path
sys.path.append('.')

# Importar módulos do projeto
from algorithms.counting_sort import counting_sort
from algorithms.radix_sort import radix_sort # Re-added for completeness, though not used in the specific benchmark loop by default
from structures.list_linear import ListLinear
from structures.list_dynamic import ListDynamic
from structures.queue_linear import QueueLinear
from structures.queue_dynamic import QueueDynamic
from structures.stack_linear import StackLinear
from structures.stack_dynamic import StackDynamic
from utils.csv_reader import load_ratings_csv, generate_test_data
from utils.data_structure import DataStructure


# --- Estruturas de Dados e Funções de Medição Refatoradas ---

@dataclass
class BenchmarkResult:
    """Resultado de um único benchmark para uma estrutura e tamanho."""
    structure_name: str
    structure_type: str
    data_size: int
    time_ms: float
    memory_mb: float

def get_memory_usage_bytes() -> float:
    """Retorna o uso de memória atual em bytes."""
    process = psutil.Process()
    return process.memory_info().rss

def force_gc():
    """Força coleta de lixo para estabilizar a medição de memória."""
    gc.collect()
    time.sleep(0.05) # Small sleep to allow GC to complete if heavy

# Function to perform the full sort cycle for time measurement
def perform_sort_cycle(ds: DataStructure, data: List[int]):
    """Performs the data insertion, conversion, sorting, and reconversion."""
    ds.from_array(data)
    arr = ds.to_array()
    sorted_arr = counting_sort(arr) # Using counting_sort as specified
    ds.from_array(sorted_arr)
    # Important: Check if sorted_arr is actually sorted and matches expected
    # This check is crucial for correctness validation within the main loop.
    return sorted_arr

def medir_tempo(ds: DataStructure, data: List[int], sort_func: Callable) -> float:
    """Executa o ciclo completo de ordenação e retorna o tempo em milissegundos."""
    # We pass the sort_func to allow flexibility if other sorts were to be used.
    # The actual sorting happens inside perform_sort_cycle.
    start_time = time.perf_counter()
    
    ds.from_array(data)
    arr = ds.to_array()
    sorted_arr = sort_func(arr) # Use the provided sort function
    ds.from_array(sorted_arr)

    end_time = time.perf_counter()
    return (end_time - start_time) * 1000  # Converte para milissegundos

def medir_memoria(ds: DataStructure, data: List[int], sort_func: Callable) -> float:
    """Executa o ciclo completo de ordenação e retorna o PICO de memória usada em bytes."""
    
    tracemalloc.start() 
    # Force GC before starting tracemalloc to clear prior allocations
    force_gc() 

    # --- Executa a mesma lógica da medição de tempo ---
    # Need to re-insert data as the structure might have been cleared or modified by time measurement.
    ds.from_array(data)
    arr = ds.to_array()
    sorted_arr = sort_func(arr) # Use the provided sort function
    ds.from_array(sorted_arr)
    # --- Fim da lógica ---

    current, peak = tracemalloc.get_traced_memory()
    
    tracemalloc.stop()
    
    # Clean up structure and force GC to minimize impact on next measurements
    ds.from_array([]) 
    force_gc()

    return peak # Retorna o PICO de memória alocada durante o bloco


def log_system_info(log_file):
    """Registra informações do sistema no log."""
    log_file.write("="*60 + "\n")
    log_file.write("     BENCHMARK - PYTHON - COUNTING SORT      \n")
    log_file.write("="*60 + "\n")
    log_file.write(f"Data/Hora: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
    log_file.write(f"Sistema: {platform.system()}/{platform.machine()}\n")
    log_file.write(f"Python: {platform.python_version()}\n")
    log_file.write("="*60 + "\n\n")


def create_summary_table(log_file, all_results: Dict[str, Dict[int, BenchmarkResult]], sizes: List[int]):
    """Cria e salva a tabela resumo final no arquivo de log."""
    structures_order = [
        "Lista Linear", "Lista Dinâmica",
        "Fila Linear", "Fila Dinâmica",
        "Pilha Linear", "Pilha Dinâmica"
    ]
    
    log_file.write("\n" + "="*80 + "\n") # Expanded width for better formatting
    log_file.write("                   TABELA RESUMO FINAL\n")
    log_file.write("="*80 + "\n\n")
    
    # Cabeçalho da tabela
    header = f"{'Estrutura':<20} | {'Tipo':<8} | "
    for size in sizes:
        header += f"{size:>10} | "
    log_file.write(header + "\n")
    
    separator = "-"*20 + "-+-" + "-"*8 + "-+-"
    for _ in sizes:
        separator += "-"*11 + "+-"
    log_file.write(separator + "\n")
    
    # Seção de Tempo
    log_file.write("\nTEMPO MÉDIO (ms):\n")
    for struct_name in structures_order:
        if struct_name in all_results:
            results_by_size = all_results[struct_name]
            # Ensure we have a result for the first size to get structure_type
            struct_type = results_by_size[sizes[0]].structure_type if sizes[0] in results_by_size else "N/A"
            line = f"{struct_name:<20} | {struct_type:<8} | "
            for size in sizes:
                if size in results_by_size:
                    line += f"{results_by_size[size].time_ms:10.2f} | "
                else:
                    line += f"{'N/A':>10} | " # If no data for this size
            log_file.write(line + "\n")

    # Seção de Memória (Pico)
    log_file.write("\nPICO DE MEMÓRIA (MB):\n")
    for struct_name in structures_order:
        if struct_name in all_results:
            results_by_size = all_results[struct_name]
            struct_type = results_by_size[sizes[0]].structure_type if sizes[0] in results_by_size else "N/A"
            line = f"{struct_name:<20} | {struct_type:<8} | "
            for size in sizes:
                if size in results_by_size:
                    line += f"{results_by_size[size].memory_mb:10.2f} | "
                else:
                    line += f"{'N/A':>10} | " # If no data for this size
            log_file.write(line + "\n")


def main():
    """Função principal que orquestra o benchmark."""
    log_filename = f"benchmark_python_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
    
    with open(log_filename, 'w', encoding='utf-8') as log_file:
        log_system_info(log_file)
        
        print("=== BENCHMARK PYTHON - COUNTING SORT ===")
        print(f"Resultados serão salvos em: {log_filename}")
        
        NUM_REPETICOES = 10 # Defined here for easier access
        sizes_to_test = [100, 1000, 10000, 100000, 1000000]
        
        all_results: Dict[str, Dict[int, BenchmarkResult]] = {}

        structures_to_test: Dict[str, Callable[[int], DataStructure]] = {
            "Lista Linear": lambda size: ListLinear(size),
            "Lista Dinâmica": lambda size: ListDynamic(),
            "Fila Linear": lambda size: QueueLinear(size),
            "Fila Dinâmica": lambda size: QueueDynamic(),
            "Pilha Linear": lambda size: StackLinear(size),
            "Pilha Dinâmica": lambda size: StackDynamic(),
        }
        
        csv_path = "/home/john/Desktop/NonComparativeSortingAlgorithms/ml-25m/ratings.csv"

        for size in sizes_to_test:
            print(f"\n--- Testando com {size} elementos ---")
            log_file.write(f"\n=== TAMANHO DO DATASET: {size} elementos ===\n")

            data = load_ratings_csv(csv_path, size)
            if not data:
                print(f"Aviso: Não foi possível ler o CSV. Gerando dados de teste para o tamanho {size}.")
                data = generate_test_data(size)
            
            if not data: # If data is still empty after trying both methods
                print(f"Erro: Não foi possível obter dados para o tamanho {size}. Pulando.")
                log_file.write(f"ERRO: Não foi possível obter dados para o tamanho {size}. Pulando.\n")
                continue

            # It's good practice to get the expected sorted data once per data set
            expected_sorted_data = sorted(data)
            
            for name, create_func in structures_to_test.items():
                ds_instance = create_func(size) # Create a fresh instance for each benchmark
                struct_type = "linear" if "Linear" in name else "dynamic"
                
                print(f"  {name:<20}: ", end='', flush=True)
                
                total_time_ms = 0.0
                successful_runs_time = 0
                first_run_memory_mb = 0.0 # Store memory from the first successful run

                # Run time measurement multiple times
                for i in range(NUM_REPETICOES):
                    # Create a new instance for each run to avoid side effects between runs
                    # particularly important for structures like queues/stacks where elements are consumed.
                    current_ds_instance_for_time = create_func(size) 
                    
                    start_time = time.perf_counter()
                    current_ds_instance_for_time.from_array(data)
                    arr_to_sort = current_ds_instance_for_time.to_array()
                    sorted_arr = counting_sort(arr_to_sort)
                    
                    # Verify sorting correctness
                    if sorted_arr == expected_sorted_data:
                        end_time = time.perf_counter()
                        total_time_ms += (end_time - start_time) * 1000
                        successful_runs_time += 1
                    else:
                        print(f"❌ Falha de ordenação na repetição {i+1} para {name} (tamanho {size}).", end="")
                        # Consider breaking here or marking as failed for this structure/size combination
                        # For now, we'll just not count this run towards the average.
                
                # Measure memory once after all time measurements, using a fresh instance
                # This gives a representative peak memory for this size/structure.
                force_gc() # Force GC before memory measurement
                ds_instance_for_mem = create_func(size)
                peak_memory_bytes = medir_memoria(ds_instance_for_mem, data, counting_sort)
                first_run_memory_mb = peak_memory_bytes / (1024.0 * 1024.0)

                if successful_runs_time > 0:
                    average_time_ms = total_time_ms / successful_runs_time
                else:
                    average_time_ms = -1.0 # Indicate failure or no successful runs

                # Store the result (average time, single memory measurement)
                result = BenchmarkResult(
                    structure_name=name,
                    structure_type=struct_type,
                    data_size=size,
                    time_ms=average_time_ms,
                    memory_mb=first_run_memory_mb
                )
                
                if name not in all_results:
                    all_results[name] = {}
                all_results[name][size] = result
                
                if average_time_ms >= 0:
                    print(f"Média (Tempo: {average_time_ms:.2f}ms, Memória Pico: {first_run_memory_mb:.2f}MB)")
                    log_file.write(f"{name} -> Tempo Médio: {average_time_ms:.4f} ms, Memória Pico: {first_run_memory_mb:.4f} MB\n")
                else:
                    print(f"Falha (Memória Pico: {first_run_memory_mb:.2f}MB)")
                    log_file.write(f"{name} -> FALHA (Tempo), Memória Pico: {first_run_memory_mb:.4f} MB\n")

        # Gerar a tabela resumo final no arquivo de log
        create_summary_table(log_file, all_results, sizes_to_test)
        
        log_file.write("\n" + "="*60 + "\n")
        log_file.write("                 BENCHMARK FINALIZADO\n")
        log_file.write("="*60 + "\n")
    
    print(f"\n✅ Benchmark concluído! Resultados salvos em: {log_filename}")


if __name__ == "__main__":
    main()